File: cpp/markov_decision_process/grid.cpp
#include "grid.h"

using namespace std;
using namespace Eigen;

Grid::Grid(int size, GridState init, GridState goal) : size(size), goal(goal)
{
    initialState = init;
}

int Grid::stateToIndex(GridState state)
{
    return get<0>(state) * size + get<1>(state);
}

GridState Grid::indexToState(int index)
{
    return GridState{index / size, index % size};
}

tuple<int, VectorXd> Grid::getRewardAndPTrans(int stateIndex, GridAction u)
{
    GridState oldState = indexToState(stateIndex);
    GridState nextState;

    VectorXd PTrans(size * size);

    if (u == GridAction::stay)
        nextState = oldState;
    else if (u == GridAction::top)
    {
        if (get<0>(oldState) == 0)
        {
            PTrans(stateIndex) = 1;
            return {-1, PTrans};
        }
        nextState = GridState{get<0>(oldState) - 1, get<1>(oldState)};
    }
    else if (u == GridAction::left)
    {
        if (get<1>(oldState) == 0)
        {
            PTrans(stateIndex) = 1;
            return {-1, PTrans};
        }
        nextState = GridState{get<0>(oldState), get<1>(oldState) - 1};
    }
    else if (u == GridAction::right)
    {
        if (get<1>(oldState) == size - 1)
        {
            PTrans(stateIndex) = 1;
            return {-1, PTrans};
        }
        nextState = GridState{get<0>(oldState), get<1>(oldState) + 1};
    }
    else if (u == GridAction::bottom)
    {
        if (get<0>(oldState) == size - 1)
        {
            PTrans(stateIndex) = 1;
            return {-1, PTrans};
        }
        nextState = GridState{get<0>(oldState) + 1, get<1>(oldState)};
    }

    if (oldState == goal)
    {
        PTrans(stateIndex) = 1;
        return {1, PTrans};
    }

    int nextStateIndex = stateToIndex(nextState);
    PTrans(nextStateIndex) = 1;
    return {0, PTrans};
}

bool Grid::isDone(GridState state)
{
    return state == goal;
}

GridState Grid::play(GridState oldState, GridAction u)
{
    GridState nextState = oldState;

    if (u == GridAction::top && get<0>(oldState) > 0)
        nextState = GridState{get<0>(oldState) - 1, get<1>(oldState)};
    else if (u == GridAction::left && get<1>(oldState) > 0)
        nextState = GridState{get<0>(oldState), get<1>(oldState) - 1};
    else if (u == GridAction::right && get<1>(oldState) < size - 1)
        nextState = GridState{get<0>(oldState), get<1>(oldState) + 1};
    else if (u == GridAction::bottom && get<0>(oldState) < size - 1)
        nextState = GridState{get<0>(oldState) + 1, get<1>(oldState)};

    return nextState;
}

File: src_py/markov_decision_process/grid.py
import numpy as np


class Grid:
    def __init__(self):
        self.action_options = ['stay', 'top', 'left', 'right', 'bottom']

        self.size = 6
        self.goal = (4, 3)

    def state_to_index(self, state):
        state_num = state[0] * self.size + state[1]
        return state_num

    def index_to_state(self, n):
        state = (n // self.size, n % self.size)
        return state

    def get_reward_P_trans(self, state_index, u):
        old_state = self.index_to_state(state_index)
        P_trans = np.zeros(self.size * self.size)

        if u == 0:
            next_state = old_state
        elif u == 1:
            if old_state[0] == 0:
                P_trans[state_index] = 1
                return -1, P_trans
            next_state = (old_state[0] - 1, old_state[1])
        elif u == 2:
            if old_state[1] == 0:
                P_trans[state_index] = 1
                return -1, P_trans
            next_state = (old_state[0], old_state[1] - 1)
        elif u == 3:
            if old_state[1] == self.size - 1:
                P_trans[state_index] = 1
                return -1, P_trans
            next_state = (old_state[0], old_state[1] + 1)
        elif u == 4:
            if old_state[0] == self.size - 1:
                P_trans[state_index] = 1
                return -1, P_trans
            next_state = (old_state[0] + 1, old_state[1])

        if old_state[0] == self.goal[0] and old_state[1] == self.goal[1]:
            P_trans[state_index] = 1
            return 1, P_trans

        next_state_index = self.state_to_index(next_state)
        P_trans[next_state_index] = 1

        return 0, P_trans

    def is_done(self, state):
        return state[0] == self.goal[0] and state[1] == self.goal[1]

    def random_play(self, state):
        u = np.random.choice(np.arange(5))
        return u, self.play(state, u)

    def play(self, old_state, u):
        next_state = old_state

        if u == 1 and old_state[0] > 0:
            next_state = (old_state[0] - 1, old_state[1])
        elif u == 2 and old_state[1] > 0:
            next_state = (old_state[0], old_state[1] - 1)
        elif u == 3 and old_state[1] < self.size - 1:
            next_state = (old_state[0], old_state[1] + 1)
        elif u == 4 and old_state[0] < self.size - 1:
            next_state = (old_state[0] + 1, old_state[1])

        return next_state
